doc/source/data/batch_inference.rst

```diff
@@ -418,3 +418,92 @@ Suppose your cluster has 4 nodes, each with 16 CPUs. To limit to at most
         compute=ray.data.ActorPoolStrategy(size=12)
         )
     predictions.show(limit=1)
+
+
+Using models from Ray Train
+---------------------------
+
+Models that have been trained with :ref:`Ray Train <train-docs>` can then be used for batch inference with :ref:`Ray Data <data>` via the :class:`Checkpoint <ray.air.checkpoint.Checkpoint>` that is returned by :ref:`Ray Train <train-docs>`.
+
+**Step 1:** Train a model with :ref:`Ray Train <train-docs>`.
+
+.. testcode::
+
+    import ray
+    from ray.train.xgboost import XGBoostTrainer
+    from ray.air.config import ScalingConfig
+
+    dataset = ray.data.read_csv("s3://anonymous@air-example-data/breast_cancer.csv")
+    train_dataset, valid_dataset = dataset.train_test_split(test_size=0.3)
+
+    trainer = XGBoostTrainer(
+        scaling_config=ScalingConfig(
+            num_workers=2,
+            use_gpu=False,
+        ),
+        label_column="target",
+        num_boost_round=20,
+        params={
+            "objective": "binary:logistic",
+            "eval_metric": ["logloss", "error"],
+        },
+        datasets={"train": train_dataset, "valid": valid_dataset},
+    )
+    result = trainer.fit()
+
+.. testoutput::
+    :hide:
+
+    ...
+
+**Step 2:** Extract the :class:`Checkpoint <ray.air.checkpoint.Checkpoint>` from the training :class:`Result <ray.air.Result>`.
+
+.. testcode::
+
+    checkpoint = result.checkpoint
+
+**Step 3:** Use Ray Data for batch inference. To load in the model from the :class:`Checkpoint <ray.air.checkpoint.Checkpoint>` inside the Python class, use one of the :ref:`framework-specific Checkpoint classes <train-framework-catalog>`.
+
+In this case, we use the :class:`XGBoostCheckpoint <ray.train.xgboost.XGBoostCheckpoint>` to load the model.
+
+The rest of the logic looks the same as in the `Quickstart <#quickstart>`_.
+
+.. testcode::
+    
+    from typing import Dict
+    import numpy as np
+    import xgboost
+
+    from ray.air import Checkpoint
+    from ray.train.xgboost import XGBoostCheckpoint
+
+    test_dataset = valid_dataset.drop_columns(["target"])
+
+    class XGBoostPredictor:
+        def __init__(self, checkpoint: Checkpoint):
+            xgboost_checkpoint = XGBoostCheckpoint.from_checkpoint(checkpoint)
+            self.model = xgboost_checkpoint.get_model()
+        
+        def __call__(self, data: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+            # Stack all the features into a single Numpy array.
+            features = np.stack(list(data.values()), axis=1)
```

Comment:
This depends on the ordering of the dict -- how do you make sure the ordering was the same during training?

Reply:
If this didn't hold, then there would lots of issues with Ray Data right now :)

Reply:
It may very well be the case, but it is very dangerous to depend on such things. We should always design things in a way that makes models not depend on implementation details like this. First, do you really understand why the dict insertion order is guaranteed? Will it not break in the future? What should a user do if they want to construct their own batch for prediction, they won't know the right insertion order, right?

Models are artifacts that should not depend on implementation details of the code. Can we improve this?

Reply:
This is a core primitive that Ray Datasets relies upon (and is guaranteed by Python). If you are saying we cannot rely on this, then Ray Datasets as a whole is not robust, and all of our examples are incorrect.

Reply:
So you are saying the ordering of the keys in the batch dict is part of the API contract? Is this documented somewhere (if not, we should add it)? Do we say where the canonical ordering is coming from (e.g. how it is mapping from the original datasource) so we can give users guidance on how they should construct the dicts if they do it manually without Ray Data.

This is very non-obvious I think btw :)

Reply:
Basically, dict ordering is guaranteed by Python. Column ordering is also guaranteed by Pyarrow. Ray Datasets depends on Python and Pyarrow. By definition, we assume that Python and Pyarrow work as intended.

Reply:
Alright, let's call this out in the docs then. We should say that the Ray Data dicts for the batches will always have the same ordering as the columns in the underlying files. Let's also make sure we go through all the possible ways how datasets can be constructed and that this is actually something we can guarantee :)

This is important, because if we do things like in this example, we need to make sure somebody can use the model without Ray Data :)

Reply:
Hmm maybe I'm misunderstanding you. Are you talking about the case where you don't use Ray Data for training, but you are using it for inference?

In this particular example, both the train set and test set are coming from the same original dataset. So therefore, the column ordering is the same all throughout.

Reply:
I'm talking about the case where you use Ray Data for the training, and then later you load the checkpoint and do the prediction with your own custom predictor class. In that case you need to know what the ordering or the features is you need to pass in :)

Much like in your example but instead of doing batch prediction with Ray Data, you just do prediction on your own, e.g. on a single example.

Reply:
To be more concrete, in this example it would be
```
predictor = XGBoostPredictor(checkpoint)
predictor.predict({"first_feature": 1, "second_feature": 2, ...})
```
People need to know that if they put the features in the right order in the dict, things will work, and that we guarantee that.

Reply:
Makes sense, though I'm not sure what needs to be changed here? We don't use `XGBoostPredictor` in the example.

Reply:
ok ok, I get what you're saying now. You're not talking about the Python implementation details, but the implementation details on how the arrow tables are inserted into the numpy dict.

I switched to pandas batch format to avoid any confusion. Generally, for this type of tabular data, pandas is probably the better option anyways.

Reply:
Yeah, that's what I meant :)

Sounds good, I think you are right pandas is clearer here. For serving probably not though -- you wouldn't want pandas on the critical path of your serving since it can have many conversion surprises with unclean data.